### PR TITLE
compute loss only if training and update token metric naming

### LINKED ISSUE #3291: `train/total_tokens` increases from eval and resets when resuming from checkpoint
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

- `train/total_tokens` should only increase from training steps, not evaluation steps.
- `train/total_tokens` should be resumed when using `resume_from_checkpoint`, not reset to 0.

### Current behaviour

- `train/total_tokens` increases during evaluation. You can see these as a jump on the step after eval. Meaning if you do evaluation every 10 steps, then step 11,21,31,41,51,etc will all show a jump. If you do evaluation every 50 steps, then step 51,101,151,201,etc will all show a jump. If you are evaluating often the final value will be very different from the true value.

These two runs use the same exact config and pre-processed dataset, but by step 50 the value is basically double the true value.
Image

- `train/total_tokens` resets to 0 when resuming.

### Steps to reproduce

Should happen with any config where evaluation is done, or the checkpoint is being resumed.

### Config yaml

```yaml

```

### Possible solution

_No response_

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12.3

### axolotl branch-commit

main/b234532

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

### PR DESCRIPTION
**Pull request by ved1beta**
fixes #3291 
image

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed token-per-second tracking to only occur when the model is actively training. Previously, tracking could run regardless of training state.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by xzuyn** [comment]
Solves the eval increase part, but the reset on resume still happens.

Eval every 10 steps.
image

**Comment by ved1beta** [comment]
ahh i thought it was just the recomputing here you goo , 
image

File: src/axolotl/utils/callbacks/tokens_per_second.py [hunk 1]
```diff
@@ -10,22 +12,50 @@
...
+        """Restore total_tokens state when resuming from checkpoint."""
+        if not isinstance(self.resume_from_checkpoint, str):
+            return
+        tokens_state_path = os.path.join(self.resume_from_checkpoint, TOKENS_STATE_FILE)
+        if os.path.isfile(tokens_state_path):
+            with open(tokens_state_path, "r", encoding="utf-8") as f:
+                tokens_state = json.load(f)
+            state.total_tokens = torch.tensor(tokens_state.get("total_tokens", 0))
+            state.num_tokens = torch.tensor(tokens_state.get("num_tokens", 0))
+            LOG.info(f"Restored total_tokens: {state.total_tokens}")
```
Comment (NanoCode012) [inline src/axolotl/utils/callbacks/tokens_per_second.py]: Is this how we should store total_tokens? Are we able to inject it into the trainer state file that gets created in checkpoints?

Comment (ved1beta) [reply src/axolotl/utils/callbacks/tokens_per_second.py]: ummm TrainerState is a dataclass, it only serializes defined fields

**Comment by xzuyn** [comment]
Resuming works too. `[2025-12-06 13:23:07,478] [INFO] [axolotl.utils.callbacks.tokens_per_second] Restored total_tokens: 544092`
image

**Review by NanoCode012 — CHANGES_REQUESTED** [review]
As mentioned in chat, let's refactor to use `tokens/total` and `tokens/trainable`. I believe `self.num_tokens` is redundant?

**Comment by ved1beta** [comment]
train eval working fine ,also checkpoints 
image

File: src/axolotl/core/trainers/base.py [hunk 2]
```diff
@@ -348,24 +351,34 @@ def compute_loss(
...
+
+            # trainable tokens for throughput and total token slots for summaries
+            self.state.tokens["trainable"] = (
+                self.state.tokens["trainable"] + trainable_tokens.detach().cpu()
+            )
+            self.state.tokens["total"] = (
+                self.state.tokens["total"] + torch.as_tensor(total_tokens).cpu()
+            )
+            # Store per-step trainable tokens for throughput calculation
+            self.state.tokens["trainable_step"] = trainable_tokens.detach().cpu()
```
Comment (NanoCode012) [inline src/axolotl/core/trainers/base.py]: Can be removed as is unused (not logged and too similar to others)

File: src/axolotl/core/trainers/base.py [hunk 3]
```diff
@@ -625,13 +638,17 @@ def log(self, logs: dict[str, float], start_time: float | None = None) -> None:
...
+            and hasattr(self.state, "tokens")
+        ):
             # each rank will log its own tokens per second
             # for logging_steps > 1 we obtain a moving average of this metric
-            logs["tokens_per_second_per_gpu"] = round(
+            logs["tokens/trainable_per_second_per_gpu"] = round(
                 self.state.last_tokens_per_second.item() / self.args.logging_steps, 2
             )
-            logs["total_tokens"] = int(self.state.total_tokens.item())
+            logs["tokens/total"] = int(self.state.tokens["total"].item())
```
Comment (NanoCode012) [inline src/axolotl/core/trainers/base.py]: I think we missed log `tokens/trainable`

File: src/axolotl/utils/callbacks/tokens_per_second.py [hunk 4]
```diff
@@ -60,5 +95,15 @@ def on_log(
...
+        tokens = getattr(state, "tokens", None)
+        # Clear per-step tokens after logging
+        if tokens and "trainable_tokens" in tokens:
+            tokens["trainable_tokens"] = torch.zeros_like(tokens["trainable_tokens"])
+
+        if tokens and "total" in tokens:
+            logs["tokens/total"] = tokens["total"].item()
+
+        if tokens and "trainable" in tokens:
+            logs["tokens/trainable"] = tokens["trainable"].item()
```
Comment (NanoCode012) [inline src/axolotl/utils/callbacks/tokens_per_second.py:109]: Is this duplicate log of base.py L651-652?

Comment (ved1beta) [reply src/axolotl/utils/callbacks/tokens_per_second.py:109]: yess. redundant ?

**Review by NanoCode012 — APPROVED** [review]

**Comment by winglian** [comment]
can we add a CI similar to TestResumeLlama that on resume the total tokens is correct?

**Comment by ved1beta** [comment]
Okay!

**Comment by xzuyn** [comment]
`tokens/total` and `tokens/trainable` aren't being sent to wandb. It is also being shown as a float instead of an int.
